test/pointing_party/vote_calculator_test.exs

```diff
@@ -0,0 +1,87 @@
+defmodule PointingParty.VoteCalculatorTest do
+  use ExUnit.Case, async: true
+  use ExUnitProperties
+
+  describe "calculate_votes/1" do
+    setup do
+      points_map = fixed_map(%{points: integer(1..5)})
+
+      metas_map =
+        fixed_map(%{
+          metas: list_of(points_map, length: 1)
+        })
+
+      user = nonempty(map_of(string(:alphanumeric), metas_map))
+      [user: user]
+    end
+
+    property "winning value is a list or a integer", %{user: user} do
+      check all(users <- user) do
+        {_event, winner} = PointingParty.VoteCalculator.calculate_votes(users)
+        assert is_list(winner) || is_integer(winner)
+      end
+    end
+
+    property "tie when winning value is a list, winner when winning value is an integer", %{
+      user: user
+    } do
+      check all(users <- user) do
+        {event, winner} = PointingParty.VoteCalculator.calculate_votes(users)
+
+        cond do
+          is_list(winner) ->
+            assert event == "tie"
+
+          is_integer(winner) ->
+            assert event == "winner"
+        end
+      end
+    end
+
+    property "the winning value is not more than the highest point value", %{user: user} do
+      check all(users <- user) do
+        {_event, winner} = PointingParty.VoteCalculator.calculate_votes(users)
+
+        max_vote =
+          users
+          |> Enum.map(fn {_username, %{metas: [%{points: points}]}} -> points end)
+          |> Enum.max()
+
+        cond do
+          is_list(winner) ->
+            winner <= max_vote
+
+          is_integer(winner) ->
+            winner <= max_vote
+        end
+      end
+    end
+
+    property "when the winner is a list of two sorted values", %{user: user} do
+      check all(users <- user) do
+        {_event, winner} = PointingParty.VoteCalculator.calculate_votes(users)
+
+        if is_list(winner) do
+          assert length(winner) == 2
+
+          votes = Enum.map(users, fn {_username, %{metas: [%{points: points}]}} -> points end)
+
+          calculated_votes =
+            Enum.reduce(votes, %{}, fn vote, acc ->
+              acc
+              |> Map.get_and_update(vote, &{&1, (&1 || 0) + 1})
+              |> elem(1)
+            end)
+
+          sorted =
+            calculated_votes
+            |> Enum.sort_by(&elem(&1, 1))
+            |> Enum.take(2)
+            |> Enum.map(&elem(&1, 0))
```

Comment:
The code from lines 69-80 is the same as the application code. That's generally not ideal because it won't catch bugs in the original algorithm. Is there a different way we can express this code in the test?

Reply:
Yea I was definitely wondering this myself! I took a stab at finding a _slightly_ different way of expressing it, but there is at least one line that is still the same as application code. Let me know what you think.